reorganize existing docs (#99)
signalfx/signalfx-java-tracing: README.md

# SignalFx Java Agent

The SignalFx Java Agent is a

> [Java Virtual Machine](https://docs.oracle.com/javase/7/docs/api/java/lang/instrument/package-summary.html)

Reviewer: this url is for the instrument package and not the jvm, so suggest "agent" be or be included as the link.

(JVM) agent that automatically instruments your Java application to capture and
report distributed traces to SignalFx. Download the JAR for the agent's 
[latest release](https://github.com/signalfx/signalfx-java-tracing/releases/latest)
and add its path to your JVM startup options:

```bash

> $ java -javaagent:path/to/signalfx-tracing.jar -jar app.jar

Reviewer: Mind adding
`curl -L https://github.com/signalfx/signalfx-java-tracing/releases/latest/download/signalfx-tracing.jar -o signalfx-tracing.jar` before the java invocation?

Reviewer: `-javaagent:./signalfx-tracing.jar` since it's downloaded in the previous step.

```
For more information, see [Configure the SignalFx Java Agent](#Configure-the-SignalFx-Java-Agent).

The agent instruments supported libraries and frameworks with bytecode
manipulation and configures an OpenTracing-compatible tracer to capture
and export trace spans. The agent also registers an OpenTracing `GlobalTracer`

> so you can add custom instrumentation to your application later.

Reviewer: An aim of this is to support existing OT custom instrumentation as well as add on if none exists.

By default, the tracer has constant sampling (i.e., 100% of spans) and
reports every span. Context propagation uses
[B3 headers](https://github.com/openzipkin/b3-propagation).

> For more information about configuring and using the agent, see

Reviewer: This document should be the most informative resource, so maybe focus on these being modifiable examples to see it in action?

| _Spark Java_ | 2.3+ | `sparkjava` | |
| Spring Data | 1.8.0+ | `spring-data` | |
| Spring Web (MVC) | 4.0+ | `spring-web` | Includes DispatcherServlet and HandlerAdapter. Supports exceptions whitelist using `@TraceSetting`. |
| Spring WebFlux | 5.0.0+ | `spring-webflux` | |
| Vertx Web | 3.0.0+  | `vertx` | This works through the Netty instrumentation for requests, and also includes spans for handlers. |

## Configure the SignalFx Java Agent

Send traces from your Java application to a local or remote Smart Agent,
OpenTelemetry Collector, or SignalFx ingest endpoint.

### Configuration values

The agent needs the following properties or environment variables for configuring
tracer functionality and trace content. System property values take priority
over corresponding environment variables.

| System property | Environment variable | Default value | Notes |
| ---             | ---                  | ---           | ---   |
| `signalfx.service.name` | `SIGNALFX_SERVICE_NAME` | `"unnamed-java-app"` | The name of the service. |
| `signalfx.agent.host` | `SIGNALFX_AGENT_HOST` | `"localhost"` | The endpoint for a SignalFx Smart Agent or OpenTelemetry Collector. |

> | `signalfx.endpoint.url` | `SIGNALFX_ENDPOINT_URL` | `"http://localhost:9080/v1/trace"` | Takes priority over constituent Agent properties. If you're sending traces directly to SignalFx, specify the SignalFx endpoint here. |

Reviewer: again, not supported atm

Author: just to be clear, remove `SIGNALFX_ENDPOINT_URL`?

Reviewer: sending to ingest

| `signalfx.trace.annotated.method.blacklist` | `SIGNALFX_TRACE_ANNOTATED_METHOD_BLACKLIST` | `null` | Prevents `@Trace` annotation functionality for the target method string of format `package.OuterClass[methodOne,methodTwo];other.package.OuterClass$InnerClass[*];`. All methods in class need to include `*` and `;`. |
| `signalfx.max.spans.per.trace` | `SIGNALFX_MAX_SPANS_PER_TRACE` | `0 (no limit)` | Drops traces with more spans than this. Intended to prevent runaway traces from flooding upstream systems. |
| `signalfx.max.continuation.depth` | `SIGNALFX_MAX_CONTINUATION_DEPTH` | `100` | Stops propagating asynchronous context at this recursive depth. Intended to prevent runaway traces from leaking memory. |

### Steps

Follow these steps to configure the agent to send traces for `your_app` to a
Smart Agent available on `localhost`. To send traces to a remote Smart
Agent, instead specify the `signalfx.agent.host` system property or
`SIGNALFX_AGENT_HOST` environment variable before you include the Java Agent in your
application.

1. Download the latest version of the agent from the
[releases](https://github.com/signalfx/signalfx-java-tracing/releases) page.

> 2. Set the required environment variables or system properties for your

Reviewer: properties are set w/ the java invocation or in properties files.  No real reason to make this a separate step since the env vars are for the java process.

Author: will the provided steps work just as well with step 2 removed?

Reviewer: yes, with the default service name.  you can also have a single instruction:
`SIGNALFX_SERVICE_NAME="your_app" java -javaagent...`

Author: going to leave in the step so users can name the service before sending data in.

### Configure the OpenTracing `GlobalTracer`

The SignalFx Java Agent configures an OpenTracing-compatible tracer
to capture and export trace spans. It registers this tracer as the OpenTracing
`GlobalTracer` to easily enable custom instrumentation throughout your
application:
```java
import io.opentracing.util.GlobalTracer;

public class MyClass {
  public void MyLogic() {
    // Use the GlobalTracer utility to create and modify an active span.
    final Tracer tracer = GlobalTracer.get();

    // Depending on MyLogic's context within a supported framework or library, the custom span
    // created here can automatically become part of an applicable agent-generated trace,
    // like those created for web framework filter activity.
    final Span span = tracer.buildSpan("MyOperation").start();
    try (Scope scope = tracer.scopeManager().activate(span)) {
        span.setTag("MyTag", "MyValue");
        <your functionality...>
    } finally {
       span.finish();
    }
  }
}
```

### Configure a `Trace` annotation

If you want to configure custom instrumentation and don't want to use the
OpenTracing `GlobalTracer` and API directly, configure a `@Trace` annotation.

> You can disable the annotation at runtime with the `signalfx.trace.annotated.method.blacklist`

Reviewer: link back to configuration table?

# SignalFx Java Agent

The SignalFx Java Agent is a
[Java Virtual Machine](https://docs.oracle.com/javase/7/docs/api/java/lang/instrument/package-summary.html)
(JVM) agent that automatically instruments your Java application to capture and
report distributed traces to SignalFx. Download the JAR for the agent's 
[latest release](https://github.com/signalfx/signalfx-java-tracing/releases/latest)

> and add it to your JVM path.

Reviewer: I wouldn't change the existing startup options language or preferably remove the `-javaagent` example at the beginning since it clearly demonstrates the simplicity of use.

Author: We can include it at the top and then link to the section for more information.

# SignalFx Java Agent

The SignalFx Java Agent is a
[Java Virtual Machine](https://docs.oracle.com/javase/7/docs/api/java/lang/instrument/package-summary.html)
(JVM) agent that automatically instruments your Java application to capture and
report distributed traces to SignalFx. Download the JAR for the agent's 
[latest release](https://github.com/signalfx/signalfx-java-tracing/releases/latest)

> and add it to your JVM path:

Reviewer: "adding its path to your JVM startup options"

```bash
$ java -javaagent:path/to/signalfx-tracing.jar -jar app.jar
```
For more information, see [Configure the SignalFx Java Agent](#Configure-the-SignalFx-Java-Agent).

The agent instruments supported libraries and frameworks with bytecode
manipulation and configures an OpenTracing-compatible tracer to capture

> and export trace spans. The agent also registers the tracer as the OpenTracing
> `GlobalTracer` so you can add custom instrumentation to your application later.

Reviewer 2: ```suggestion
and export trace spans. The agent also registers an OpenTracing
`GlobalTracer` so you can add custom instrumentation to your application later.
```

> By default, the tracer has constant sampling (i.e., 100% chance of tracing) and
> reports every span. Where applicable, context propagation uses
> [B3 headers](https://github.com/openzipkin/b3-propagation).

Reviewer 2: ```suggestion
By default, the tracer has constant sampling (i.e., 100% of spans).
In addition, [B3](https://github.com/openzipkin/b3-propagation)
context propagation is used.
```

For more information about configuring and using the agent, see
the [examples](https://github.com/signalfx/tracing-examples/tree/master/signalfx-tracing/signalfx-java-tracing).

## Requirements and supported software

> These are the requirements and supported software for the agent.
> 

Reviewer 2: ```suggestion
```

Specify the SignalFx Java Agent as the only JVM agent for your application.
If you specify multiple agents, you may encounter issues with at least one

> of them.

Reviewer 2: This is a repeat of above -- remove. In addition, the warning should be above.

Author: I don't follow. What should we remove?

The agent works with Java runtimes version 8 and higher. Other JVM-based
languages like Scala and Kotlin are also supported, but may not work with all
instrumentations.

These are the supported libraries. _Italicized_ libraries are in beta. Enable

> beta libraries by setting the `-Dsignalfx.integration.<name>.enabled=true` 

Reviewer 2: nit: command wraps on page, should all be on same line

system property, where `<name>` is the instrumentation name specified in
the table.

> | Library | Versions supported | Instrumentation name(s) | Notes |

Reviewer 2: Table is wider than page, consider single instrumentation name per line

| Memcached (SpyMemcached) | 2.10.0+ | `spymemcached` | |
| Mongo Client | 3.1+ | `mongo` | |
| Mongo Async Client | 3.3+ | `mongo` | |
| Netty Client and Server | 4.0+ | `netty` | Nonstandard HTTP status code tagging w/ `-Dsignalfx.instrumentation.netty.{client,server}.nonstandard.http.status.<code>=true` to circumvent Status5xxDecorator. |
| OkHTTP Client | 3.0+ | `okhttp` | |
| Play Web Framework | 2.4+ | `play` | |
| RabbitMQ Client | 2.7.0+ | `rabbitmq` | |
| Ratpack | 1.4+ | `ratpack` | |
| Reactor Core | 3.1.0+ | `reactor-core` | |
| RestTemplate | 3.1.1+ | `rest-template` | |
| Java Servlet | 2+ | `servlet` | |
| _Spark Java_ | 2.3+ | `sparkjava` | |
| Spring Data | 1.8.0+ | `spring-data` | |
| Spring Web (MVC) | 4.0+ | `spring-web` | Includes DispatcherServlet and HandlerAdapter. Supports exceptions whitelist using `@TraceSetting`. |
| Spring WebFlux | 5.0.0+ | `spring-webflux` | |
| Vertx Web | 3.0.0+  | `vertx` | This works through the Netty instrumentation for requests, and also includes spans for handlers. |

## Configure the SignalFx Java Agent

Send traces from your Java application to a local or remote Smart Agent,

> OpenTelemetry Collector, or SignalFx ingest endpoint.

Reviewer 2: Each option should link to product docs

Author: Should we remove OpenTelemetry Collector from this since it's not part of the happy path?

Reviewer: sending directly to ingest isn't supported since there's no access token option.  OTel collector is the happy path.

| `signalfx.db.statement.max.length` | `SIGNALFX_DB_STATEMENT_MAX_LENGTH` | `1024` | The maximum number of characters written for the OpenTracing `db.statement` tag. |
| `signalfx.recorded.value.max.length` | `SIGNALFX_RECORDED_VALUE_MAX_LENGTH` | `12288` | The maximum number of characters for any Zipkin-encoded tagged or logged value. |
| `signalfx.trace.annotated.method.blacklist` | `SIGNALFX_TRACE_ANNOTATED_METHOD_BLACKLIST` | `null` | Prevents `@Trace` annotation functionality for the target method string of format `package.OuterClass[methodOne,methodTwo];other.package.OuterClass$InnerClass[*];`. All methods in class need to include `*` and `;`. |
| `signalfx.max.spans.per.trace` | `SIGNALFX_MAX_SPANS_PER_TRACE` | `0 (no limit)` | Drops traces with more spans than this. Intended to prevent runaway traces from flooding upstream systems. |
| `signalfx.max.continuation.depth` | `SIGNALFX_MAX_CONTINUATION_DEPTH` | `100` | Stops propagating asynchronous context at this recursive depth. Intended to prevent runaway traces from leaking memory. |

### Steps

> To send traces to a local or remote Smart Agent or OpenTelemetry Collector, specify

Reviewer 2: We should only document the happy path -- you could send directly to Otel Collector, but this is not the recommendation. Just talk about sending to Smart Agent

Author: Smart Agent and ingest endpoint, or just Smart Agent?

> the `signalfx.agent.host` system property or `SIGNALFX_AGENT_HOST` environment

Reviewer 2: This is only needed if the defaults do not work -- ideally, nothing should need to be specified and it should work out of the box. The only time this is not the case may be a containerized environment where `localhost` cannot be reached.

variable before you include the agent in your application. If you send traces
to a Smart Agent, specify a `v1` endpoint. If you send traces to an
OpenTelemetry Collector, specify a `v2` endpoint.

> To send traces to a SignalFx ingest endpoint, specify the `signalfx.endpoint.url`

Reviewer 2: This should only be recommended for lambdas when an otel collector cannot be deployed. Otherwise, again should recommend only to smart agent. We have a separate repo for lambda so would recommend removing from here.

system property or `SIGNALFX_ENDPOINT_URL` environment variable before you
include the agent in your application. If you're using µAPM PM, specify a `v1`
endpoint in the URL. If you're using the latest µAPM, specify a `v2` endpoint
in the URL.

Follow these steps to configure the agent to send traces for `your_app` to a
local Smart Agent available on `localhost`.

1. Download the latest version of the agent from the
[releases](https://github.com/signalfx/signalfx-java-tracing/releases) page.
2. Set the required environment variables or system properties for your
application. For more information about the required environment variables,
see the [Configuration values](#configuration-values).
This is how you can set the proper environment variables from the command line:
    ```bash
    $ export SIGNALFX_SERVICE_NAME="your_app"
    $ export SIGNALFX_AGENT_HOST="localhost"
    $ export SIGNALFX_TRACING_ENABLED="true"
    $ export SIGNALFX_SPAN_TAGS="key1:val1,key2:val2"
    $ export SIGNALFX_DB_STATEMENT_MAX_LENGTH=1024
    $ export SIGNALFX_RECORDED_VALUE_MAX_LENGTH=12288
    $ export SIGNALFX_TRACE_ANNOTATED_METHOD_BLACKLIST=null
    $ export SIGNALFX_MAX_SPANS_PER_TRACE=0

>     $ export SIGNALFX_MAX_CONTINUATION_DEPTH=100

Reviewer 2: For the happy path, none of these should need to be specified.

Author: Should we leave span tags in for environment tag or anything else?

    ```
1. Include the agent in your Java application: 
    ```bash
    $ java -javaagent:path/to/signalfx-tracing.jar -jar app.jar
    ```

## Troubleshoot the SignalFx Java Agent

Enable debug logging for troubleshooting assistance. Set the
`datadog.slf4j.simpleLogger.defaultLogLevel=debug` property at runtime.  These logs

> are extremely verbose.

Reviewer 2: Should have a note about how this should only be enabled when needed and immediately turned off -- it will impact performance on the application.

## Manually instrument a Java application

> You can use the OpenTracing `GlobalTracer` or a `Trace` annotation to manually

Reviewer 2: ```suggestion
You can use the OpenTracing `GlobalTracer` or a `@Trace` annotation to manually
```

instrument your Java application.

### Configure the OpenTracing `GlobalTracer`

The SignalFx Java Agent configures an OpenTracing-compatible tracer
to capture and export trace spans. It registers this tracer as the OpenTracing
`GlobalTracer` to easily enable custom instrumentation throughout your
application:
```java

> import io.opentracing.util.GlobalTracer;

Reviewer 2: @rmfitzpatrick is this correct?

Reviewer: Yes, but `import io.opentracing.*;` is also needed for this snippet for Tracer, Scope, and Span.

| `signalfx.max.spans.per.trace` | `SIGNALFX_MAX_SPANS_PER_TRACE` | `0 (no limit)` | Drops traces with more spans than this. Intended to prevent runaway traces from flooding upstream systems. |
| `signalfx.max.continuation.depth` | `SIGNALFX_MAX_CONTINUATION_DEPTH` | `100` | Stops propagating asynchronous context at this recursive depth. Intended to prevent runaway traces from leaking memory. |

### Steps

Follow these steps to configure the agent to send traces for `your_app` to a
Smart Agent available on `localhost`. To send traces to a remote Smart
Agent, instead specify the `signalfx.agent.host` system property or
`SIGNALFX_AGENT_HOST` environment variable before you include the Java Agent in your
application.

1. Download the latest version of the agent from the
[releases](https://github.com/signalfx/signalfx-java-tracing/releases) page.
2. Set the required environment variables or system properties for your
application. For more information about the required environment variables,
see the [Configuration values](#configuration-values).
Set these environment variables from the command line:
    ```bash
    $ export SIGNALFX_SERVICE_NAME="your_app"

>     $ export SIGNALFX_SPAN_TAGS="key1:val1,key2:val2"

Reviewer 3: SPAN_TAGS is not required and shouldn't be in this section.  ENDPOINT, maybe.

    ```
1. Include the agent in your Java application: 
    ```bash
    $ java -javaagent:path/to/signalfx-tracing.jar -jar app.jar
    ```

## Troubleshoot the SignalFx Java Agent

Enable debug logging for troubleshooting assistance. Set the

> `datadog.slf4j.simpleLogger.defaultLogLevel=debug` property at runtime. These

Author: Would it be better to change the title to "Enable debug logging"

> `datadog.slf4j.simpleLogger.defaultLogLevel=debug` property at runtime. These

Reviewer 3: It's fine enough as it is.  A search for "debug" on the page brings you right to this point.

    // This error will not cause the span to have error tag
    throw new InvalidArgumentException();
}
```

### Track span context across threads

> Use the Java Agent to track span context across thread boundaries. Provide

Reviewer: "boundaries, assuming your asynchronous or concurrent workers are supported."  If not this has no effect.

explicit marker for spans to automatically propagate them when using Java's
standard concurrency tools.  If you already have access to the current scope
(e.g., from an ``activate()`` call), set the `async propagation` flag on the
span like this:

```java
import com.signalfx.tracing.context.TraceScope;

// ...

    Span span = GlobalTracer.get().buildSpan("my-operation").start();
    try (Scope sc = GlobalTracer.get().scopeManager().activate(span, true)) {
        // The below cast will always work as long as you haven't set a custom tracer.
        ((TraceScope) scope).setAsyncPropagation(true);
        // ... Dispatch work to a Java thread.
        // Any methods calls in the new thread will have their active scope set to the current one.
    }
```

If you don't have access to the scope that determines whether the operation
should be continued across threads, you can get it from the `GlobalTracer`
like this:

```java
import com.signalfx.tracing.context.TraceScope;

// ...

    Span span = GlobalTracer.get().buildSpan("my-operation").start();

>     try (Scope sc = GlobalTracer.get().scopeManager().activate(span, true)) {

Reviewer: The point of this example to show that they haven't created their own span/scope or have an immediate handle on it.  Only the inner line is intended (as I think the original example was written).

Author: what exactly needs to change here?

Reviewer: remove lines 268-269,274
